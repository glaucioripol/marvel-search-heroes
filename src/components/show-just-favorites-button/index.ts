export { ShowJustFavoritesButton } from "./show-just-favorites-button";
