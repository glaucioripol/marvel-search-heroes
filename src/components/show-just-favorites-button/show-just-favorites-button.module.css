.just_liked_button {
  background-color: transparent;
  border: none;
  border-radius: 32px;
  cursor: pointer;
  padding: 8px 16px;
}

.just_liked_button:hover {
  background-color: var(--red-1);
  outline: 2px dashed var(--red-3);
}

.just_liked_button.just_liked_button--active {
  background-color: var(--red-1);
}

.just_liked_button.just_liked_button--disabled {
  background-color: var(--grey-4);
  color: #fff;
  cursor: not-allowed;
}

.just_liked_button.just_liked_button--disabled:hover {
  outline: 2px dashed var(--grey-3);
}
