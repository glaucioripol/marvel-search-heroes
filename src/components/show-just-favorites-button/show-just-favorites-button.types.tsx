export type ShowJustFavoritesButtonProperties = Readonly<{
  isActive: boolean;
  onClick?: (value: boolean) => void;
  dataTestId?: string;
  disabled?: boolean;
}>;
