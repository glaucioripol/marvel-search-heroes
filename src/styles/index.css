:root {
  --default-font: "Roboto", sans-serif;

  /* Colors */
  --red-1: #fdecec;
  --red-2: #fa7c7c;
  --red-3: #fa8787;
  --red-4: #ff0b0b;
  --red-5: #ff1510;
  --red-6: #e92429;

  --grey-1: #404040; /* usando: Search heroes ao lado da logo */
  --grey-2: #464646; /* usando: titulo acima do texto */
  --grey-3: #969696; /* usando: texto acima do input */
  --grey-4: #bdbdbd; /* usando: Número de dados encontrados */

  /* Background colors */
  --background-color: #ffffff;
  --footer-background-color: var(--red-5);

  /* Font sizes */
  /* texto página herói: Quadrinhos e Filmes  */
  --font-x-sm: 18px;
  /* texto padrão */
  --font-sm: 20px;
  /* Descrição do herói */
  --font-md: 24px;
  /* Titulo página inicial */
  --font-xxl: 44px;
  /* Titulo página do herói, input home */
  --font-xl: 36px;
}

/* reset css */
* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

body {
  background-color: var(--background-color);
  font-family: var(--default-font);
}
