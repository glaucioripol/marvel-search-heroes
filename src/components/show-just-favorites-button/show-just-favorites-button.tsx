import styles from "./show-just-favorites-button.module.css";

import { KeyboardEvent, MouseEvent, useCallback, useMemo } from "react";

import { ShowJustFavoritesButtonProperties } from "./show-just-favorites-button.types";

export function ShowJustFavoritesButton({
  isActive,
  onClick,
  dataTestId = "just-liked-button",
  disabled = false,
}: ShowJustFavoritesButtonProperties) {
  const currentClassName = useMemo(
    () =>
      [
        styles.just_liked_button,
        "text_with_icon",
        isActive ? styles["just_liked_button--active"] : "",
        disabled ? styles["just_liked_button--disabled"] : "",
      ].join(" "),
    [disabled, isActive],
  );

  const handleOnClick = useCallback(
    (event: MouseEvent<HTMLButtonElement>) => {
      event.preventDefault();
      event.stopPropagation();

      onClick?.(!isActive);
    },
    [isActive, onClick],
  );

  const handleKeyDown = useCallback(
    (event: KeyboardEvent<HTMLButtonElement>) => {
      event.preventDefault();
      event.stopPropagation();

      const eventsToListen = [" ", "Enter"];

      if (eventsToListen.includes(event.key)) {
        onClick?.(!isActive);
      }
    },
    [isActive, onClick],
  );

  return (
    <button
      type="button"
      onClick={handleOnClick}
      onKeyDown={handleKeyDown}
      data-testid={dataTestId}
      aria-label="Exibir somente favoritos"
      title="Exibir somente favoritos"
      className={currentClassName}
      disabled={disabled}>
      <img
        src="/assets/favorito_01.svg"
        alt="Imagem de um coração, representando um favorito."
        width={18}
        height={18}
      />
      <span>Somente favoritos</span>
    </button>
  );
}
