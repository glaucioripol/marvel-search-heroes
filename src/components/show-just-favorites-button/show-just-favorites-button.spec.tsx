import {
  defaultRender,
  describe,
  expect,
  fireEvent,
  it,
  vi,
  screen,
  beforeEach,
} from "@/__tests__";
import { ShowJustFavoritesButton } from "./show-just-favorites-button";
import styles from "./show-just-favorites-button.module.css";

describe("ShowJustFavoritesButton", () => {
  let onClickMock = vi.fn();

  beforeEach(() => {
    onClickMock = vi.fn();
  });

  const renderSUT = (isActive: boolean, disabled = false) => {
    const utils = defaultRender(
      <ShowJustFavoritesButton
        isActive={isActive}
        onClick={onClickMock}
        disabled={disabled}
      />,
    );

    const button = screen.getByTestId("just-liked-button");

    return { ...utils, button };
  };

  it("renders without crashing", () => {
    renderSUT(false);
    expect(screen.getByTestId("just-liked-button")).toBeInTheDocument();
  });

  it("renders with correct initial state", () => {
    const { button, asFragment } = renderSUT(false);
    expect(button).toHaveAttribute("aria-label", "Exibir somente favoritos");
    expect(button).toHaveAttribute("title", "Exibir somente favoritos");
    expect(button).toHaveClass(styles.just_liked_button);
    expect(asFragment()).toMatchSnapshot();
  });

  it("calls onClick when clicked", () => {
    const { button } = renderSUT(false);
    fireEvent.click(button);
    expect(onClickMock).toHaveBeenCalledWith(true);
  });

  it("toggles state correctly on repeated clicks", () => {
    const { button } = renderSUT(true);
    fireEvent.click(button);
    expect(onClickMock).toHaveBeenCalledWith(false);
  });

  it("changes attributes and class when clicked", () => {
    const { button, rerender, asFragment } = renderSUT(false);

    fireEvent.click(button);
    rerender(<ShowJustFavoritesButton isActive={true} onClick={onClickMock} />);

    expect(button).toHaveClass(styles["just_liked_button--active"]);
    expect(asFragment()).toMatchSnapshot();
  });

  it("disables button when disabled prop is true", () => {
    const { button } = renderSUT(false, true);
    expect(button).toBeDisabled();
    expect(button).toHaveClass(styles["just_liked_button--disabled"]);
  });

  describe("keyboard interactions", () => {
    it("calls onClick on space key press", () => {
      const { button } = renderSUT(false);
      fireEvent.keyDown(button, { key: " " });
      expect(onClickMock).toHaveBeenCalledWith(true);
    });

    it("calls onClick on Enter key press", () => {
      const { button } = renderSUT(false);
      fireEvent.keyDown(button, { key: "Enter" });
      expect(onClickMock).toHaveBeenCalledWith(true);
    });

    it("does not call onClick on other key press", () => {
      const { button } = renderSUT(false);
      fireEvent.keyDown(button, { key: "a" });
      expect(onClickMock).not.toHaveBeenCalled();
    });
  });
});
