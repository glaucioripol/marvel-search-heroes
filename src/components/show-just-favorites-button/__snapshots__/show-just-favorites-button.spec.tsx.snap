// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ShowJustFavoritesButton > changes attributes and class when clicked 1`] = `
<DocumentFragment>
  <button
    aria-label="Exibir somente favoritos"
    class="_just_liked_button_62679c text_with_icon _just_liked_button--active_62679c "
    data-testid="just-liked-button"
    title="Exibir somente favoritos"
    type="button"
  >
    <img
      alt="Imagem de um coração, representando um favorito."
      height="18"
      src="/assets/favorito_01.svg"
      width="18"
    />
    <span>
      Somente favoritos
    </span>
  </button>
</DocumentFragment>
`;

exports[`ShowJustFavoritesButton > renders with correct initial state 1`] = `
<DocumentFragment>
  <button
    aria-label="Exibir somente favoritos"
    class="_just_liked_button_62679c text_with_icon  "
    data-testid="just-liked-button"
    title="Exibir somente favoritos"
    type="button"
  >
    <img
      alt="Imagem de um coração, representando um favorito."
      height="18"
      src="/assets/favorito_01.svg"
      width="18"
    />
    <span>
      Somente favoritos
    </span>
  </button>
</DocumentFragment>
`;
